src/dataset.rs

```diff
@@ -56,36 +59,64 @@ impl<T: Number, U: Number> Dataset<T, U> {
     }
 
     /// Return all of the indices in the dataset.
-    pub fn indices(&self) -> Indices { (0..self.data.shape()[0]).collect() }
+    pub fn indices(&self) -> Indices {
+        (0..self.data.shape()[0]).collect()
+    }
 
     /// Returns the number of rows in the dataset.
-    pub fn nrows(&self) -> usize { self.data.nrows() }
+    pub fn nrows(&self) -> usize {
+        self.data.nrows()
+    }
 
     /// Returns the shape of the dataset.
-    pub fn shape(&self) -> &[usize] { self.data.shape() }
+    pub fn shape(&self) -> &[usize] {
+        self.data.shape()
+    }
 
     /// Clears the internal cache.
-    pub fn clear_cache(&self) { self.cache.clear() }
+    pub fn clear_cache(&self) {
+        self.cache.clear()
+    }
 
     /// Returns the size of the internal cache.
-    pub fn cache_size(&self) -> usize { self.cache.len() }
+    pub fn cache_size(&self) -> usize {
+        self.cache.len()
+    }
 
     /// Return a random selection of unique indices.
+    /// 
+    /// Returns `n` unique random indices from the provided vector.
+    /// If `n` is greater than the number of indices provided, the full list is returned in shuffled order.
     #[allow(clippy::ptr_arg)]
     pub fn choose_unique(&self, indices: Indices, n: usize) -> Indices {
         // TODO: actually check for uniqueness among choices
-        indices.into_iter().choose_multiple(&mut rand::thread_rng(), n)
+        let mut x = indices;
+        x.shuffle(&mut rand::thread_rng());
+        x.truncate(n);
+        x
```

Comment:
This is functionally identical to the old version. choose_multiple already chooses without replacement. We need to make certain that no pair of indices correspond to points whose distance apart is zero.